### DISCUSSION TITLE
Docker Operator TLS CERTIFICATE_VERIFY_FAILED

### DISCUSSION THREAD

**Discussion by shawnbro** [apache/airflow#16645] [General]
I'm trying to use the DockerOperator but I am  running into a TLS error. I've reproduced the error in python with the following code (which i pulled from the DockerOperator source):

```
import docker
tls_config = docker.tls.TLSConfig(ca_cert='./ca.pem', client_cert=('./cert.pem', './key.pem'), verify=True)
client = docker.APIClient(base_url='https://****.com:2375', tls=tls_config)
client.version()
```

which returns this error:
```
requests.exceptions.SSLError: HTTPSConnectionPool(host='****.com', port=2375): Max retries exceeded with url: /v1.35/version (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get issuer certificate (_ssl.c:1123)')))
```

This is a certificate which was issued from cloudflare, so it's not self-signed.  When I curl my remote docker provider with the same configuration, it verifies the certificate authority just fine.

`curl --cert cert.pem --key key.pem https://****:2375/version` returns
```
{"Platform":{"Name":""},"Components":[{"Name":"Engine","Version":"20.10.4","Details":{"ApiVersion":"1.41","Arch":"amd64","BuildTime":"2021-03-29T18:55:03.000000000+00:00","Experimental":"false","GitCommit":"363e9a8","GoVersion":"go1.15.8","KernelVersion":"4.14.232-177.418.amzn2.x86_64","MinAPIVersion":"1.12","Os":"linux"}},{"Name":"containerd","Version":"1.4.4","Details":{"GitCommit":"05f951a3781f4f2c1911b05e61c160e9c30eaa8e"}},{"Name":"runc","Version":"1.0.0-rc93","Details":{"GitCommit":"12644e614e25b05da6fd08a38ffa0cfe1903fdec"}},{"Name":"docker-init","Version":"0.19.0","Details":{"GitCommit":"de40ad0"}}],"Version":"20.10.4","ApiVersion":"1.41","MinAPIVersion":"1.12","GitCommit":"363e9a8","GoVersion":"go1.15.8","Os":"linux","Arch":"amd64","KernelVersion":"4.14.232-177.418.amzn2.x86_64","BuildTime":"2021-03-29T18:55:03.000000000+00:00"}
```

Any idea what I'm doing wrong here? Here is my DockerOperator code in the Dag.

```
t1 = DockerOperator(
    dag=dag,
    task_id='docker_task',
    environment=env_dict,
    image=image,
    auto_remove=True,
    api_version='auto',
    docker_url='tcp://***.com:2375',
    docker_conn_id='docker_default',
    command='python app.py',
    tls_ca_cert=f'{os.getcwd()}/ca.pem',
    tls_client_cert=f'{os.getcwd()}/cert.pem',
    tls_client_key=f'{os.getcwd()}/key.pem',
)
```

and those files exist on the airflow machine.

Thanks,

Shawn

**Comment by potiuk** [comment]
How about the cacert ? What do you have there? I f you are providing cacert then it overrides the built in CA certs and only uses the CA's that are specified in you cacert file. So unless you add cloudflare's CA to the file it will not validate.

**Reply by shawnbro** [reply]
I think I am providing the ca cert from cloud flare. Unless I’m mistaken.

**Reply by potiuk** [reply]
Try not to provide one ?. The 'curl' command you mentioned as working uses default cacert apparently.

**Reply by shawnbro** [reply]
the issue with that is that the DockerOperator source code expects all three to be there, otherwise it won't use TLS altogether.

**Reply by potiuk** [reply]
Aargh. Then I think you might not have the whole signing chain there. I think when you provide the cacert, you have to provide thw whole certificate chain in the cacert file. Lilkely you can test it with curl adding the same file via `--with-ca-bundle` - this will let you verify if your cacert file is good.

**Reply by shawnbro** [reply]
👍 will give that a shot and get back to you, thank you for your input.
